database.c

```diff
@@ -312,12 +312,9 @@ void save_to_DB(void)
 	for(i = lastdbindex; i < counters.queries; i++)
 	{
 		validate_access("queries", i, true, __LINE__, __FUNCTION__, __FILE__);
-		if(queries[i].timestamp <= lasttimestamp || queries[i].db == true)
-		{
-			// Already in database
-			// logg("Skipping %i",i);
+		if(queries[i].timestamp <= lasttimestamp || queries[i].db || !queries[i].complete)
+			// Already in database or not yet complete
 			continue;
```

Comment:
If we skip a query while it's still not yet complete and the one after is complete, we would lose the first query. Perhaps change this to a `break` so that we never add any queries after one which is not complete.

Reply:
Hmm, not sure if that is a good solution - although it never happend to me so far, this could be a potential dead-lock assuming that `dnsmasq` didn't write out the result for one query so it would never be completed. What do you think?

Reply:
Maybe bring back the 200 line check or convert it to a timestamp check. If the query hasn't completed after 2 seconds, assume it's a goner and continue.

Reply:
Okay, I will in addition compare the time stamp of the query with the current time and `break` if the query was from within the most recent two seconds. It will have a chance to get saved (or eventually ignored) on the next DB dumping time.